Add peak routing example

TheGreatRefrigerator opened the pull request:
in this example the routes to different mountain peaks,
extracted from OSM with overpy, are calculated for three
starting points.
The resulting maps are transformed to PNGs using
selenium and webdriver-manager.
Optional image editing is done by example of a resize
using Pillow.

This example was contributed by Dr. Stephan Fuchs

Co-authored-by: Stephan <nicefuchs@gmail.com>

TheGreatRefrigerator committed:
Add peak routing example

in this example the routes to different mountain peaks,
extracted from OSM with overpy, are calculated for three
starting points.
The resulting maps are transformed to PNGs using
selenium and webdriver-manager.
Optional image editing is done by example of a resize
using Pillow.

This example was contributed by Dr. Stephan Fuchs

Co-authored-by: Stephan <nicefuchs@gmail.com>

koebi commented on python/peak_routing_png_image.py:
```diff
@@ -0,0 +1,218 @@
+from webdriver_manager.chrome import ChromeDriverManager
+import time
+
+chrome_driver = webdriver.Chrome(ChromeDriverManager().install())
```
For this, chrome has to be installed.
For firefox, use 
```
from selenium import webdriver
from webdriver_manager.firefox import GeckoDriverManager

driver = webdriver.Firefox(executable_path=GeckoDriverManager().install())
```
Maybe link to https://pypi.org/project/webdriver-manager/ from above for people that have neither chrome nor ff installed?

TheGreatRefrigerator replied:
will do

TheGreatRefrigerator replied:
done

koebi commented on python/peak_routing_png_image.py:
```diff
@@ -0,0 +1,218 @@
+
+# Query natural peaks from overpass api within bounding box
+result = api.query(
+    f"""node["natural"="peak"][name~"^({"|".join(peak_names)})$"]({",".join([str(c) for c in bbox])});out;""")
```
Why is this using a multiline string? if you need double quotes in the query, use single quotes for the string? or is my python bad here?

Also, this line is utterly unreadable. is there a nice wrapper around querys, to make it more clear what is going on?
Maybe one could comment the query above, so people understand?

TheGreatRefrigerator replied:
1. valid point.. guess i alt-enterd for auto completion, single quote works as well.
2. that's the overpass query language for you, writing a whole wrapper for that would be out of scope for an example i guess..

what about the comment right above it that tells that it queries natural peaks from overpass api within the bounding box, you think that's not sufficient?
there is not much more info i can add except for saying that it will query point objects where the name matches exactly one of the strings

koebi replied:
yeah,  that's the comment in natural language.
I was thinking along the lines of having the query there without all the f"{}" and python-functions in there, so people familiar with overpass can read it or copy it into overpass to understand?

TheGreatRefrigerator replied:
Ah i see. I was thinking in the direction of making the values easily interchangable if one wants to run the example on other bbox & peaks.
maybe a print of the query before sending it to make it possible to copy from the cell output?

TheGreatRefrigerator replied:
Did a print of the output to easily see the whole query

koebi replied:
Nice, I like the approach and the usability :)

koebi commented on python/peak_routing_png_image.py:
```diff
@@ -0,0 +1,218 @@
+# For each of the peaks we create a folium map with routes from every starting point, and add it to the `maps` list.
+# To every map we add the points, the calculated routes and a little styling.
+#
+# __Make sure to insert a valid openrouteservice API key before running the cell__
```
I like this comment :) Was rather valuable to me.

koebi commented on python/peak_routing_png_image.py:
```diff
@@ -0,0 +1,218 @@
+
+    # calculate overview-bounds
+    bbox = [
+        [min(c[0][0] for c in bounds),
```
this looks like it could be written a bit more concise, given that there are 4 list comprehensions here…

TheGreatRefrigerator replied:
there every bound is a list of `[[minx,miny],[maxx, maxy]]` and you need to filter by the value.
This was the most concise i could come up with...
Suggestions welcome.

koebi commented on python/peak_routing_png_image.py:
```diff
@@ -0,0 +1,218 @@
+# -
+
+# #### Save maps as PNG
+# For transformation to PNG, we need a browser to take a screenshot from.
```
this seems a bit weird - is there no other way to export png from folium/leaflet?
there seems to be a workaround using PIL: https://stackoverflow.com/questions/53565979/export-a-folium-map-as-a-png

koebi replied:
That would drop selenium and webdriver_manager from the requirements list?

TheGreatRefrigerator replied:
I did quite a bit of research on alternatives, but didn't find an other way.
Using private functions should not be a solution -.- but i will give it a try.
Not sure if the resolution can be adjusted though?
The good thing with the browser approach is you can create really high resolution maps, e.g. for posters

TheGreatRefrigerator replied:
Nope, this does not work properly.
1. highly experimental and not finished (therefore still private method)
2. also uses selenium internally which needs a webdriver, which fails if `geckodriver` is not in path.
3. no real means to adjust the resolution, which is quite handy

I would keep the webdriver_manager approach

koebi replied:
sgtm, thanks for investigating :)

TheGreatRefrigerator committed:
Updated peak routing with suggestions

- added environment note and requirements section
- reworked overpass section with full query output
- switched webdriver to firefox with possibility of chrome
- added link to further supported browsers
- added concluding sentence
- moved imports to top

TheGreatRefrigerator commented:
rebased on master

koebi commented on python/peak_routing_png_image.ipynb:
```diff
@@ -0,0 +1,555 @@
+      "    ]\n",
+      "  },\n",
+      "  {\n",
+      "    \"name\": \"Kristin Hr\\u00e1dek\",\n",
```
Just noted this weird encoding in the output - is this fixable?

TheGreatRefrigerator replied:
indeed -> done

koebi reviewed (changes requested):
It feels a bit weird, that the big "this is where we use ors and query directions and stuff"-codeblock doesn't give any feedback on success/failure - maybe one could add a print line or two :)

TheGreatRefrigerator committed:
Add utf-8 json decoding function

also:
- add more output
- add function docs

TheGreatRefrigerator committed:
Update .gitignore

- add geockodriver.log
- fix ignoring jupyter symlink

koebi reviewed (approved):
lgtm :))
